climada/engine/uncertainty/base.py

```diff
@@ -0,0 +1,771 @@
+        a numeric value, called the sensitivity index. Sensitivity indices
+        come in several forms:
+
+        First-order indices: measures the contribution to the output variance
+        by a single model input alone.
+        Second-order indices: measures the contribution to the output variance
+        caused by the interaction of two model inputs.
+        Total-order index: measures the contribution to the output variance
+        caused by a model input, including both its first-order effects
+        (the input varying alone) and all higher-order interactions.
+
+
+        Parameters
+        ----------
+        salib_method : str
+            sensitivity analysis method from SALib.analyse
+            Possible choices:
+                'fast', 'rbd_fact', 'morris', 'sobol', 'delta', 'ff'
+            The default is 'sobol'.
+            Note that in Salib, sampling methods and sensitivity analysis
+            methods should be used in specific pairs.
+            https://salib.readthedocs.io/en/latest/api.html
+        method_kwargs: dict(), optional
+            Keyword arguments of the chosen SALib analyse method.
+            The default is to use SALib's default arguments.
+        Returns
+        -------
+        sensitivity_dict : dict
+            Dictionnary of the sensitivity indices. Keys are the
+            metrics names, values the sensitivity indices dictionnary
+            as returned by SALib.
+
+        """
+        
+        check_salib(self.sampling_method, salib_method)
+        
+        #To import a submodule from a module use 'from_list' necessary
+        #c.f. https://stackoverflow.com/questions/2724260/why-does-pythons-import-require-fromlist
+        method = getattr(
+            __import__('SALib.analyze',
+                       fromlist=[salib_method]
+                       ),
+            salib_method
+            )
+        
+        #Certaint Salib method required model input (X) and output (Y), others
+        #need only ouput (Y)
+        salib_kwargs = method.analyze.__code__.co_varnames
+        X = self.samples_df.to_numpy() if 'X' in salib_kwargs else None
+
+        if method_kwargs is None: method_kwargs = {} 
+        sensitivity_dict = {}
+        for name, df_metric in self.metrics.items():
+            sensitivity_dict[name] = {}
+            for metric in df_metric:
+                Y = df_metric[metric].to_numpy()
+                if X is not None:
+                    sensitivity_index = method.analyze(self.problem, X, Y,
+                                                            **method_kwargs)
+                else:
+                    sensitivity_index = method.analyze(self.problem, Y,
+                                                            **method_kwargs)
+                sensitivity_dict[name].update({metric: sensitivity_index})
+        
+        self.sensitivity = sensitivity_dict
+
+        return sensitivity_dict
```

Comment:
From the tutorial it seems the output dict of such function is of the form (for e.g., impact):
`
{'aai_agg': {'aai_agg': sensitivity_indexes_dict},
 'freq_curve': {'rp50': sensitivity_indexes_dict, 'rp100': sensitivity_indexes_dict, 
'rp250': sensitivity_indexes_dict},
 'tot_value': {'tot_value': sensitivity_indexes_dict}}
`
why not something like:
`
{'aai_agg': sensitivity_indexes_dict,
 'rp50': sensitivity_indexes_dict, 'rp100': sensitivity_indexes_dict, 
'rp250': sensitivity_indexes_dict, 'tot_value': sensitivity_indexes_dict}
`
I think it is much easier to ready and, to me, there is no need to group the RPs under the key freq_curve, especially if this requires redundant info for all other keys.

Reply:
Good point. The reason I did it this way is for simplicity and compatibility with the cost benefit part. Basically, the keys now are directly the keys that come out of the `impact.calc` and `costbenefit.calc`. These are dictionnaries. In the case of impact, it looks a bit awkward, but for the cost benefit it makes sense to me. 

For example, `cost_benefit.calc()` produces the dictionnaries 
```
        imp_meas_future (dict): impact of each measure at future or default.
            Key: measure name ('no measure' used for case without measure),
            Value: dict with:
                             'cost' (tuple): (cost measure, cost factor insurance),
                             'risk' (float): risk measurement,
                             'risk_transf' (float): annual expected risk transfer,
                             'efc'  (ImpactFreqCurve): impact exceedance freq
                (optional)   'impact' (Impact): impact instance
        imp_meas_present (dict): impact of each measure at present.
            Key: measure name ('no measure' used for case without measure),
            Value: dict with:
                             'cost' (tuple): (cost measure, cost factor insurance),
                             'risk' (float): risk measurement,
                             'risk_transf' (float): annual expected risk transfer,
                             'efc'  (ImpactFreqCurve): impact exceedance freq
                (optional)   'impact' (Impact): impact instance
```

Both of these have the same keys, twice! Hence, I found it simpler to have nested dictionaries that directly take the outputs of the `.calc`.